returnn/torch/engine.py

```diff
@@ -515,13 +515,20 @@ def eval_model(self, *, skip_already_evaluated: bool = False):
             if skip_already_evaluated and self._is_dataset_evaluated(name=dataset_name):
                 continue
 
-            data_loader = self._eval_dataloaders[dataset_name]
+            # We need to make sure the data loader iterator was created for proper synchronization.
+            data_loader = iter(self._eval_dataloaders[dataset_name])
```

Comment:
Why did you change this? This also is a bug now. You are potentially creating the data loader twice for the main rank, but only once for the other ranks. Just keep the original code.
```suggestion
            data_loader = self._eval_dataloaders[dataset_name]
```

Reply:
Wouldn't `iter(iter(X)) == iter(X)`, i.e. repeated application be an identity?

Reply:
Usually but not necessarily. It depends what `__iter__` returns, and then what would `__iter__().__iter__()` return.

But anyway, don't change this part of the code when this was already working.

Reply:
Hm, although, maybe you are still right though, not with your question, but with the logic. I think `for ... in data_loader` will check if it is already iterable (by checking whether it has `__next__`), and if so, it would use it directly, and not call `iter` on it anymore.

But anyway, I think I would just leave this part of the code as it is now.